File: Games/blackjack.py
import random

# Global Variables
cards = []
houseCards = []
playerCards = []

# Create the deck with values
def deckManager():
    global cards
    cards.clear()
    values = ["ace", 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10]  # 10 is repeated for Jack, Queen, King
    for _ in range(4):  # 4 suits
        cards.extend(values)

# Deal cards to player and house
def cardDeal():
    global houseCards, playerCards
    houseCards.clear()
    playerCards.clear()
    
    for _ in range(2):  # Deal 2 cards to player and house
        houseCards.append(random.choice(cards))
        cards.remove(houseCards[-1])
        playerCards.append(random.choice(cards))
        cards.remove(playerCards[-1])
    
    print("Dealer has X and " + str(houseCards[1]))
    print("You have " + str(playerCards[0]) + " " + str(playerCards[1]))
    print("Your card value: " + str(calculateValue(playerCards)) + "\n")

# Calculate the value of a hand
def calculateValue(hand):
    value = 0
    ace_count = 0
    for card in hand:
        if card == "ace":
            value += 11
            ace_count += 1
        else:
            value += card
    while value > 21 and ace_count > 0:
        value -= 10  # Adjust for aces as 1 instead of 11
        ace_count -= 1
    return value

# Player's decision to hit or stand
def playHit():
    while True and calculateValue(playerCards) < 21:
        playerAction = input("Hit or Stand ?: ").strip().upper()
        if playerAction == "HIT" or playerAction == "H":
            new_card = random.choice(cards)
            cards.remove(new_card)
            playerCards.append(new_card)
            print(f"You got: {new_card}")
            print(f"Your Cards: {playerCards} \nYour Card Value: {calculateValue(playerCards)}")
        elif playerAction == "STAND" or playerAction == "S":
            break
        else:
            print("Invalid choice. Please enter 'Hit' or 'Stand'.")

# Check for Blackjack
def blackJackCheck():
    if calculateValue(playerCards) == 21:
        print("Blackjack! YOU WIN!")
    else:
        print("Dealer's turn...")

# Dealer's turn
def dealerHit():
    while calculateValue(houseCards) < 17:  # Dealer hits until the value is 17 or more
        new_card = random.choice(cards)
        cards.remove(new_card)
        houseCards.append(new_card)
        print(f"Dealer got: {new_card}")
    dealer_value = calculateValue(houseCards)
    print(f"Dealer's card value: {dealer_value}")
    
    if dealer_value > 21:
        print("Dealer Busts! You Win!")
    elif dealer_value >= calculateValue(playerCards):
        print(f"Dealer Wins! Dealer's cards: {houseCards}")
    else:
        print(f"You Win! Dealer's cards: {houseCards}")

# Game loop
while True:
    playerWantsToPlay = input("Play Blackjack? y/n : ").strip().upper()
    if playerWantsToPlay == "Y":
        print("Shuffling Deck...")
        deckManager()
        print(f"Cards remaining: {len(cards)}")
        
        # Start the game
        cardDeal()
        
        if calculateValue(playerCards) == 21:
            blackJackCheck()
        else:
            playHit()
            if calculateValue(playerCards) <= 21:
                dealerHit()
            else:
                print("You Lose! It's a bust! \n")
    else:
        print("Game Over \nThank you for playing <3")
        break

File: Games/blackjack.java
import java.util.ArrayList;
import java.util.Scanner;

public class Blackjack {

    private static final ArrayList<Object> cards = new ArrayList<>();
    private static final ArrayList<Object> houseCards = new ArrayList<>();
    private static final ArrayList<Object> playerCards = new ArrayList<>();
    private static final Scanner scanner = new Scanner(System.in);

    // Deck Manager
    private static void deckManager() {
        cards.clear();
        for (int count = 1; count <= 32; count++) {
            cards.add("ace");
            cards.add(10);
            cards.add(10);
            cards.add(10);
            for (int n = 2; n <= 10; n++) {
                cards.add(n);
            }
        }
    }

    // Deal cards to player and dealer
    private static void cardDeal() {
        houseCards.clear();
        playerCards.clear();
        for (int deal = 1; deal <= 2; deal++) {
            houseCards.add(drawCard());
            playerCards.add(drawCard());
        }
        System.out.println("Dealer has X and " + houseCards.get(1));
        handleAce(houseCards);
        System.out.println("You have " + playerCards.get(0) + " and " + playerCards.get(1));
        handleAce(playerCards);
        System.out.println("Your card value: " + calculateSum(playerCards) + "\n");
    }

    private static Object drawCard() {
        int randomIndex = (int) (Math.random() * cards.size());
        return cards.remove(randomIndex);
    }

    private static void handleAce(ArrayList<Object> hand) {
        if (hand.contains("ace")) {
            hand.remove("ace");
            hand.add(11);
        }
    }

    // Player chooses hit or stand
    private static void playHit() {
        while (calculateSum(playerCards) < 21) {
            System.out.print("Hit or Stand?: ");
            String action = scanner.nextLine();
            if (action.equalsIgnoreCase("hit") || action.equalsIgnoreCase("h")) {
                Object tempCard = drawCard();
                if (tempCard.equals("ace")) {
                    tempCard = 1;
                }
                playerCards.add(tempCard);
                System.out.println("You got: " + tempCard);
                System.out.println("Your cards: " + playerCards);
                System.out.println("Your card value: " + calculateSum(playerCards));
            } else {
                break;
            }
        }
    }

    // BlackJack check
    private static void blackJackCheck() {
        if (calculateSum(playerCards) == calculateSum(houseCards)) {
            System.out.println("Push! It's a DRAW.");
        } else {
            System.out.println("Blackjack! YOU WIN!");
            System.out.println("Dealer had: " + houseCards);
        }
    }

    // Dealer's turn
    private static void dealerHit() {
        int playerScore = calculateSum(playerCards);
        while (calculateSum(houseCards) < playerScore) {
            Object randomCard = drawCard();
            if (randomCard.equals("ace")) {
                randomCard = 1;
            }
            houseCards.add(randomCard);
            System.out.println("Dealer got: " + randomCard);
        }
        System.out.println("Dealer total card value: " + calculateSum(houseCards));
        if (calculateSum(houseCards) > playerScore && calculateSum(houseCards) <= 21) {
            System.out.println("House Wins! " + houseCards);
        } else {
            System.out.println("You Win!");
            System.out.println("Dealer Bust! His cards: " + houseCards);
        }
    }

    // Calculate hand value
    private static int calculateSum(ArrayList<Object> hand) {
        int sum = 0;
        for (Object card : hand) {
            if (card instanceof Integer) {
                sum += (int) card;
            } else if (card.equals("ace")) {
                sum += 11;
            }
        }
        return sum;
    }

    // Main game logic
    public static void main(String[] args) {
        while (true) {
            System.out.print("Play BlackJack? y/n: ");
            String playResponse = scanner.nextLine();
            if (playResponse.equalsIgnoreCase("y")) {
                System.out.println("Shuffling Deck...");
                deckManager();
                System.out.println("Deck size: " + cards.size()); //Test
            } else {
                System.out.println("Game Over \nThank you for playing <3");
                break;
            }

            // Game Logic
            cardDeal();
            if (calculateSum(playerCards) == 21) {
                blackJackCheck();
            } else {
                playHit();
                if (calculateSum(playerCards) < 21) {
                    dealerHit();
                } else {
                    System.out.println("You Lose! It's a bust!");
                }
            }
        }
    }
}
